collector/receiver/otelarrowreceiver/internal/arrow/arrow.go

```diff
@@ -573,3 +732,36 @@ func (r *Receiver) processRecords(ctx context.Context, method string, arrowConsu
 		return ErrUnrecognizedPayload
 	}
 }
+
+func (r *Receiver) acquireAdditionalBytes(ctx context.Context, uncompSize int64, response batchResp, sizeHeaderFound bool) error {
+	diff := uncompSize - response.bytesToRelease
+
+	var err error
+	if diff != 0 {
+		if sizeHeaderFound {
+			// a mismatch between header set by exporter and the uncompSize just calculated.
+			r.telemetry.Logger.Debug("mismatch between uncompressed size in receiver and otlp-pdata-size header", zap.Int("uncompsize", int(uncompSize)), zap.Int("otlp-pdata-size", int(response.bytesToRelease)))
+		} else if diff < 0 {
+
+			// proto.Size() on compressed request was greater than pdata uncompressed size.
+			r.telemetry.Logger.Debug("uncompressed size is less than compressed size", zap.Int("uncompressed", int(uncompSize)), zap.Int("compressed", int(response.bytesToRelease)))
+		}
+
+		if diff > 0 {
+			// diff > 0 means we previously acquired too few bytes initially and we need to correct this. Release previously
+			// acquired bytes to prevent deadlock and reacquire the uncompressed size we just calculated.
+			// Note: No need to release and reacquire bytes if diff < 0 because this has less impact and no reason to potentially block
+			// a request that is in flight by reacquiring the correct size.
+			r.boundedQueue.Release(response.bytesToRelease)
+			err = r.boundedQueue.Acquire(ctx, uncompSize)
+			if err != nil {
+				response.bytesToRelease = int64(0)
+			} else {
+				response.bytesToRelease = uncompSize
```

Comment:
This looks better than before, thanks for working on it.

Reply:
Thanks for the helpful review!